ert/data/_record.py

```diff
@@ -110,20 +131,34 @@ def record_type(self) -> RecordType:
             f"Not able to deduce record type from data was: {type(self.data)}"
         )
 
-    @root_validator(skip_on_failure=True)
-    def ensure_consistent_index(cls, record: Dict[str, Any]) -> Dict[str, Any]:
-        assert (
-            "data" in record and "index" in record
-        ), "both data and index must be defined for a record"
-        norm_record_index = _build_record_index(record["data"])
-        assert (
-            norm_record_index == record["index"]
-        ), f"inconsistent index {norm_record_index} vs {record['index']}"
-        return record
+
+class BlobRecord(Record, _DataElement):
+    data: blob_record_data
+
+    @property
+    def record_type(self) -> RecordType:
+        if isinstance(self.data, list) and isinstance(self.data[0], bytes):
+            return RecordType.LIST_BYTES
+        raise TypeError(
+            f"Not able to deduce record type from data was: {type(self.data)}"
+        )
+
+
+def make_record(data: Any, index: Optional[RecordIndex] = None) -> Record:
```

Comment:
The reason I did this is that `Record` is now an abstract class. Even if it was not abstract, it would not have a meaningful constructor to make all types of records, like the old `Record` class did. This factory function makes the appropriate record given an input of the right type, using the logic of pydantic. I am not very happy about the way it works, using exception handling, but that allows us to leave the data conversion to pydantic. Also, the error handling is gruesome, but I guess one would run into that in any case at some point when converting something unknown to the appropriate concreate class.

On the pro side, there are some cases in the code where records are made without referring to the type. Before that used `Record`, which could be simply replaced by `make_record`. The question is, is that feature really needed, or can we live with always explicitly making the right concrete type? We can easily remove the `make_record` function, at the cost of some more code in several places, where a helper function like this would be useful.

One reason this whole thing feels a bit wrong is that pydantic does not play that well with the concept of a base class. The type hinting is even worse, it basically breaks down in the combination pydantic/base class/mypy. You see also that there is basically no functionality in the base class. One could just have two classes, with a factory function, and everything would work the same. In conclusion: we are not benefitting much from having a base class here so far...

Reply:
As it turns out, it is not a big deal to remove the global `make_record`. It is still used as a local helper in `_record.py`.

Reply:
I think we can assume to have a great deal of knowledge of the data upon construction of the record. It is later when it is passed around we want to avoid having to deal with the different types of records...

The promotion of data through the types specified in the old `Record` was more a product of rubbing `pydantic` the right way and getting a quick start, than something we really wanted.

I think removing the public `make_record` sounds like a good idea 👍🏻